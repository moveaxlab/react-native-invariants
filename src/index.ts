export interface ReactTestRendererJSON {
  type: string;
  props: { [propName: string]: any };
  children: null | ReactTestRendererNode[];
}

type ReactTestRendererNode = ReactTestRendererJSON | string;

export function checkInvariants(tree: ReactTestRendererJSON | null) {
  if (!tree || typeof tree !== 'object') return;
  if (tree.children) {
    for (const children of tree.children) {
      if (tree.type === 'Text' && typeof children !== 'string' && children.type === 'View') {
        throw new Error('Nesting of <View> within <Text> is not currently supported.');
      } else if (tree.type !== 'Text' && typeof children === 'string') {
        throw new Error('Text strings must be rendered within a <Text> component');
      } else if (typeof children === 'object') {
        checkInvariants(children);
      }
    }
  }
}
