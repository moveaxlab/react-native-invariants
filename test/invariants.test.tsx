import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ButtonWithinText } from './examples/ButtonWithinText';
import { RawTextWithoutText } from './examples/RawTextWithoutText';
import { ViewWithinText } from './examples/ViewWithinText';
import { CorrectExample } from './examples/CorrectExample';
import { checkInvariants } from '../src';

describe('Test error examples', () => {
  [ButtonWithinText, RawTextWithoutText, ViewWithinText].forEach((Component) => {
    it(`renders ${Component.displayName} without errors`, () => {
      const output = renderer.create(<Component />).toJSON();
      expect(output).toMatchSnapshot();
    });

    it(`invariants fail while rendering ${Component.displayName}`, () => {
      const output = renderer.create(<Component />).toJSON();
      expect(output).not.toBeNull();
      expect(() => checkInvariants(output)).toThrow();
    });
  });
});

describe('Test correct example', () => {
  [CorrectExample].forEach((Component) => {
    it(`renders ${Component.displayName} without errors`, () => {
      const output = renderer.create(<Component />).toJSON();
      expect(output).toMatchSnapshot();
    });

    it(`invariants pass while rendering ${Component.displayName}`, () => {
      const output = renderer.create(<Component />).toJSON();
      expect(output).not.toBeNull();
      expect(() => checkInvariants(output)).not.toThrow();
    });
  });
});
