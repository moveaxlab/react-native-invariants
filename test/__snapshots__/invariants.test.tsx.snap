// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test correct example renders CorrectExample without errors 1`] = `
<View>
  <Text>
    This is some text
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        A button
      </Text>
    </View>
  </View>
</View>
`;

exports[`Test error examples renders ButtonWithinText without errors 1`] = `
<Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Test
      </Text>
    </View>
  </View>
</Text>
`;

exports[`Test error examples renders RawTextWithoutText without errors 1`] = `
<View>
  I cannot do this
</View>
`;

exports[`Test error examples renders ViewWithinText without errors 1`] = `
<Text>
  <View>
    I cannot do this
  </View>
</Text>
`;
